libevm/VMCalls.cpp

```diff
@@ -248,27 +250,26 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
 	ON_OP();
 	updateIOGas();
 
-	if (m_OP != Instruction::DELEGATECALL && m_SP[2] > 0)
+	if (haveValueArg && m_SP[2] > 0)
 		callParams->gas += m_schedule->callStipend;
 
 	callParams->codeAddress = destinationAddr;
 
-	unsigned inOutOffset = 0;
-	if (m_OP == Instruction::DELEGATECALL || m_OP == Instruction::STATICCALL)
+	if (haveValueArg)
 	{
-		callParams->apparentValue = m_ext->value;
-		callParams->valueTransfer = 0;
+		callParams->valueTransfer = m_SP[2];
+		callParams->apparentValue = m_SP[2];
 	}
 	else
 	{
-		callParams->apparentValue = callParams->valueTransfer = m_SP[2];
-		inOutOffset = 1;
+		callParams->apparentValue = m_ext->value;
```

Comment:
I will add tests.

1. call transaction into A with non-zero value; A performs STATICCALL to B; B stores the call value
2. call transaction into Q; Q calls into A with non-zero value; A performs STATICCALL to B; B stores the call value

In both cases, the storage slot of `B` would initially contain some non-zero value, so that we can be sure that the value is overwritten by zero.